src/iNxM.h

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright 2016 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef WABT_INXM_H_
+#define WABT_INXM_H_
+
+#include <array>
+#include <cassert>
+#include <cstdint>
+#include <climits>
+
+namespace iNxM {
+
+/// Array containing `Length` `IntBits`-bit integers
+template <uint8_t IntBits, ptrdiff_t Length>
+struct iNxM {
+  /// 64-bit integers are used to represent parsed natural numbers;
+  using text_t = uint64_t;
+  /// Internally the iN integers are stored in a byte:
+  using bin_t = uint8_t;
+  static_assert(
+    IntBits <= sizeof(bin_t) * CHAR_BIT,
+    "bin_t is not large enough to fit iN"
+  );
+
+  /// Largest value that an N-bit integer can have: [0, 2^N)
+  static constexpr text_t max_value() {
+    return (1 << IntBits) - 1;
+  }
+
+  /// Array length
+  static constexpr ptrdiff_t size() {
+    return Length;
+  }
+
+  /// In the textual encoding we treat the integers as Nat.
+  using TextValue = std::array<text_t, Length>;
+
+  static_assert(IntBits * Length % (sizeof(bin_t) * CHAR_BIT) == 0,
+                "The array length in bits must be a multiple of CHAR_BIT");
```

Comment:
This error message should be updated. For now the array has to precisely fit its layout, but we can choose `bin_t` programatically in the future if it doesn't fit.